util/histogram_windowing.cc

```diff
@@ -0,0 +1,193 @@
+}
+
+HistogramWindowingImpl::HistogramWindowingImpl() {
+  env_ = Env::Default();
+  window_stats_.reset(new HistogramStat[num_windows_]);
+  Clear();
+}
+
+HistogramWindowingImpl::HistogramWindowingImpl(
+    uint64_t num_windows,
+    uint64_t micros_per_window,
+    uint64_t min_num_per_window) :
+      num_windows_(num_windows),
+      micros_per_window_(micros_per_window),
+      min_num_per_window_(min_num_per_window) {
+  env_ = Env::Default();
+  window_stats_.reset(new HistogramStat[num_windows_]);
+  Clear();
+}
+
+HistogramWindowingImpl::~HistogramWindowingImpl(){
+  window_stats_.release();
+}
+
+void HistogramWindowingImpl::Clear() {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  stats_.Clear();
+  for (size_t i = 0; i < num_windows_; i++) {
+    window_stats_[i].Clear();
+  }
+  current_window_.store(0, std::memory_order_relaxed);
+  last_swap_time_.store(env_->NowMicros(), std::memory_order_relaxed);
+}
+
+bool HistogramWindowingImpl::Empty() const { return stats_.Empty(); }
+
+// This function is designed to be lock free, as it's in the critical path
+// of any operation.
+// Each individual value is atomic, it is just that some samples can go
+// in the older bucket which is tolerable.
+void HistogramWindowingImpl::Add(uint64_t value){
+  TimerTick();
+
+  // Parent (global) member update
+  stats_.Add(value);
+
+  // Current window update
+  window_stats_[current_window()].Add(value);
+}
+
+void HistogramWindowingImpl::Merge(const Histogram& other) {
+  if (strcmp(Name(), other.Name()) == 0) {
+    Merge(dynamic_cast<const HistogramWindowingImpl&>(other));
+  }
+}
+
+void HistogramWindowingImpl::Merge(const HistogramWindowingImpl& other) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  stats_.Merge(other.stats_);
+
+  if (stats_.num_buckets_ != other.stats_.num_buckets_ ||
+      micros_per_window_ != other.micros_per_window_) {
+    return;
+  }
+
+  uint64_t cur_window = current_window();
+  uint64_t other_cur_window = other.current_window();
+  // going backwards for alignment
+  for (unsigned int i = 0; 
+                    i < std::min(num_windows_, other.num_windows_); i++) {
+    uint64_t window_index =
+        (cur_window + num_windows_ - i) % num_windows_;
+    uint64_t other_window_index =
+        (other_cur_window + other.num_windows_ - i) % other.num_windows_;
+
+    window_stats_[window_index].Merge(other.window_stats_[other_window_index]);
+  }
+}
+
+std::string HistogramWindowingImpl::ToString() const {
+  return stats_.ToString();
+}
+
+double HistogramWindowingImpl::Median() const {
+  return Percentile(50.0);
+}
+
+double HistogramWindowingImpl::Percentile(double p) const {
+  // Retry 3 times in total
+  for (int retry = 0; retry < 3; retry++) {
+    uint64_t start_num = stats_.num();
+    double result = stats_.Percentile(p);
+    // Detect if swap buckets or Clear() was called during calculation
+    if (stats_.num() >= start_num) {
+      return result;
+    }
+  }
+  return 0.0;
+}
```

Comment:
Why don't we simply acquire lock here? I don't think Percentile() is called frequently.

Saying that, I'm fine with this best efforts implementation to avoid inconsistent reports, as I don't think keeping it consistent in this corner case is very critical. But it's your call.